scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/context/TaskContextVariableExtractor.java

```diff
@@ -0,0 +1,144 @@
+        ForkEnvironment forkEnvironment = taskContext.getInitializer().getForkEnvironment();
+        Map<String, Serializable> variables = this.extractTaskVariables(taskContext);
+        Map<String, String> thirdPartyCredentials = forkedTaskVariablesManager.extractThirdPartyCredentials(taskContext);
+        HashMap<String, Serializable> systemEnvironmentVariables = new HashMap<String, Serializable>(
+                System.getenv());
+        systemEnvironmentVariables.putAll(variables);
+        systemEnvironmentVariables.putAll(thirdPartyCredentials);
+
+        return VariableSubstitutor.filterAndUpdate(forkEnvironment.getSystemEnvironment(),
+                systemEnvironmentVariables);
+    }
+
+    public Map<String, Serializable> extractTaskVariables(TaskContext taskContext) throws Exception {
+        return extractTaskVariables(taskContext, (TaskResult) null);
+    }
+
+    private Map<String, Serializable> extractTaskVariables(TaskContext container, TaskResult taskResult,
+            String nodesFile)
+            throws Exception {
+        Map<String, Serializable> variables = extractTaskVariables(container, taskResult);
+
+        variables.put(SchedulerVars.PA_NODESNUMBER.toString(), container.getOtherNodesURLs().size() + 1);
+        variables.put(SchedulerVars.PA_NODESFILE.toString(), nodesFile);
+
+        variables.put(SchedulerVars.PA_TASK_PROGRESS_FILE.toString(), container.getProgressFilePath());
+
+        return variables;
+    }
+
+    public Map<String, Serializable> extractTaskVariables(TaskContext taskContext,
+            TaskResult taskResult) throws Exception {
+        Map<String, Serializable> variables = new HashMap<>();
+
+        // variables from workflow definition
+        if (taskContext.getInitializer().getVariables() != null) {
+            variables.putAll(taskContext.getInitializer().getVariables());
+        }
+
+        try {
+            // variables from previous tasks
+            if (taskContext.getPreviousTasksResults() != null) {
+                variables.putAll(extractPrevoiusTaskResultVariablesFromTaskContext(taskContext));
+            }
+            // and from this task execution
+            if (taskResult != null && taskResult.getPropagatedVariables() != null) {
+                variables.putAll(SerializationUtil.deserializeVariableMap(taskResult
+                        .getPropagatedVariables()));
+            }
+        } catch (Exception e) {
+            throw new Exception("Could not deserialize variables", e);
+        }
+
+        // variables from current job/task context
+        variables.putAll(retrieveContextVariables(taskContext.getInitializer()));
+
+        variables.put(SchedulerVars.PA_SCHEDULER_HOME.toString(), taskContext.getSchedulerHome());
+        return variables;
+    }
+
+
+    public Map<String, Serializable> extractTaskVariables(TaskContext container,
+            String nodesFile) throws Exception {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)